config_explorer/src/config_explorer/capacity_planner.py

```diff
@@ -130,32 +198,125 @@ def kv_cache_req(model_info: ModelInfo,
                         max_model_len: int,
-                        available_gpu_count: int,
                         gpu_memory: int,
+                        gpu_mem_util: float=0.9,
+                        tp: int=1,
+                        pp: int=1,
+                        dp: int=1,
                     ) -> int:
+
+    # Find allocatable memory for KV cache
+    gpu_counts = gpus_required(tp, pp, dp)
+    avail_gpu_mem = available_gpu_memory(gpu_memory, gpu_mem_util) * gpu_counts
+    model_memory = model_memory_req(model_info) * dp
+    kv_cache_allocatable_memory = avail_gpu_mem - model_memory
+
+    # Find kv cache requirement for one request of max-model-len
+    per_request_kv_cache_req = kv_cache_req(model_info, model_config, max_model_len)
+    return max(0, math.floor(kv_cache_allocatable_memory / per_request_kv_cache_req))
+
+def find_possible_tp(model_config: AutoConfig) -> List[int]:
+    """
+    Finds possible values for tp for the given model
+    """
+    num_attention_heads = model_config.num_attention_heads
+
+    factors = set(reduce(
+        list.__add__,
+        ([i, num_attention_heads // i] for i in range(1, int(num_attention_heads**0.5) + 1) if num_attention_heads % i == 0)))
+
+    factors = list(factors)
+    factors.sort()
+    return factors
+
+def available_gpu_memory(memory: int, gpu_utilization: float=0.9) -> float:
     """
-    Calculates the max number of concurrent requests the model can serve with the specified GPUs available
+    Returns the available GPU memory
+    """
+
+    return memory * gpu_utilization
+
+def gpus_required(tp: int, pp: int, dp: int) -> int:
+    """
+    Determines the number of GPUs required based on parallelism strategies
+    """
+
+    return tp * pp * dp
+
+def per_gpu_model_memory_required(model_info: ModelInfo, tp: int = 1, pp: int = 1) -> int:
+    """
+    Calculates model memory requirement for each GPU
     """
 
     model_memory = model_memory_req(model_info)
-    if model_memory == -1:
-        return -1
-    per_request_kv_cache = kv_cache_req(model_info,
+    return model_memory / (tp * pp)
+
+def per_gpu_memory_required(model_info: ModelInfo,
+                        model_config: AutoConfig,
+                        max_model_len: int,
+                        max_concurrency: int,
+                        tp: int = 1,
+                        pp: int = 1,
+                        ) -> int:
+    """
+    Determines the minimum per-GPU memory requirement for loading the model and serving the max concurrent request
+    """
+
+    per_gpu_model_mem = per_gpu_model_memory_required(model_info, tp, pp)
+    per_request_kv_cache_memory = kv_cache_req(model_info,
                                         model_config,
                                         max_model_len,
-                                        )
+                                        max_concurrency)
+    per_request_kv_cache_memory_per_gpu = per_request_kv_cache_memory / (tp * pp)
```

Comment:
With this formula, the interpretation of batch_size / concurrency is that it is the concurrency-per-DP-rank (or "instance") -- is it not?

If you agree, it might be a good idea to clarify in the streamlit documentation.